src/client/runner/automation/playback/scroll.js

```diff
@@ -1,364 +1,185 @@
-        right:  scrollableDimensions.right - scrollableDimensions.border.right - scrollableDimensions.scrollbar.right -
-                targetDimensions.right
-    };
-
-    // vertical scroll
-    if (relation.top < 0)
-        scrollTop = Math.round(scrollableDimensions.scroll.top + relation.top - maxTopScrollMargin);
-    else if (relation.top > 0 && relation.bottom < 0) {
-        scrollTop    = Math.round(scrollableDimensions.scroll.top + Math.min(relation.top, -relation.bottom) +
-                                  maxTopScrollMargin);
-        isScrollDown = 1;
-    }
 
-    // horizontal scroll
-    if (relation.left < 0)
-        scrollLeft = Math.round(scrollableDimensions.scroll.left + relation.left - maxLeftScrollMargin);
-    else if (relation.left > 0 && relation.right < 0) {
-        scrollLeft    = Math.round(scrollableDimensions.scroll.left + Math.min(relation.left, -relation.right) +
-                                   maxLeftScrollMargin);
-        isScrollRight = 1;
-    }
 
-    //NOTE: we should check: can we show the full element
-    var targetElementWidth  = Math.abs(targetDimensions.right - targetDimensions.left),
-        targetElementHeight = Math.abs(targetDimensions.bottom - targetDimensions.top);
+// Setup cross-iframe interaction
+messageSandbox.on(messageSandbox.SERVICE_MSG_RECEIVED_EVENT, e => {
+    if (e.message.cmd === 'scroll') {
+        var element = domUtils.getIframeByWindow(e.source);
+        var offsetX = e.message.offsetX;
+        var offsetY = e.message.offsetY;
+        var scroll  = new Scroll(element, { offsetX, offsetY });
 
-    if (scrollableDimensions.width <= targetElementWidth &&
-        offsetX >= scrollableDimensions.width - (scrollableDimensions.border.left + scrollableDimensions.border.right))
-        scrollLeft = (scrollLeft === null ? offsetX : scrollLeft + offsetX - isScrollRight * maxLeftScrollMargin) -
-                     maxLeftScrollMargin;
-
-    if (scrollableDimensions.height <= targetElementHeight &&
-        offsetY >= scrollableDimensions.height - (scrollableDimensions.border.top + scrollableDimensions.border.bottom))
-        scrollTop = (scrollTop === null ? offsetY : scrollTop + offsetY - isScrollDown * maxTopScrollMargin) -
-                    maxTopScrollMargin;
+        scroll
+            .run()
+            .then(() => {
+                messageSandbox.sendServiceMsg({ cmd: 'scroll-response' }, e.source);
+            });
+    }
+});
 
-    if (scrollLeft !== null || scrollTop !== null) {
-        return {
-            top:  scrollTop,
-            left: scrollLeft
-        };
+export default class Scroll {
+    constructor (element, { offsetX, offsetY }) {
+        this.element = element;
+        this.offsetX = offsetX;
+        this.offsetY = offsetY;
     }
 
-    return null;
-}
+    static _setScroll (element, { left, top }) {
+        var isDocumentElement = /html/i.test(element.tagName);
```

Comment:
no, we have not. We have `isDocument` but it's the check for Document instance, not for DOM element